lib/broadway/consumer.ex

```diff
@@ -56,30 +54,13 @@ defmodule Broadway.Consumer do
     {:noreply, [], state}
   end
 
-  defp ack_messages(successful_messages, failed_messages) do
-    %{}
-    |> group_by_acknowledger(successful_messages, :successful)
-    |> group_by_acknowledger(failed_messages, :failed)
-    |> Enum.each(&call_ack/1)
-  end
-
-  defp group_by_acknowledger(grouped_messages, messages, key) do
-    Enum.reduce(messages, grouped_messages, fn %{acknowledger: {acknowledger, _}} = msg, acc ->
-      Map.update(acc, acknowledger, [{key, msg}], &[{key, msg} | &1])
-    end)
-  end
+  defp handle_batch(publisher_key, messages, batch_info, state) do
+    %{module: module, context: context} = state
 
-  defp call_ack({acknowledger, messages}) do
-    {successful, failed} = unpack_messages(messages, [], [])
-    acknowledger.ack(successful, failed)
+    try do
+      module.handle_batch(publisher_key, messages, batch_info, context)
+    rescue
+      e -> Enum.map(messages, &Message.failed(&1, e))
```

Comment:
I would prefer if the reason was `{:failed, string}`  so we know that the reason is always a certain type. I think the failure handling because of an exception is going to be different to one caused by the user called `Message.failed` because we don't have to worry about it for now. My suggestion for now is to change this to:

    message = Exception.message(e)
    Enum.map(messages, &Message.failed(&1, message))